bftengine/src/bcstatetransfer/RangeValidationTree.hpp

```diff
@@ -0,0 +1,268 @@
+// LICENSE file.
+
+#pragma once
+
+#include <iostream>
+#include <vector>
+#include <memory>
+#include <unordered_map>
+#include <string>
+#include <cmath>
+#include <limits>
+
+#include <cryptopp/integer.h>
+
+#include "STDigest.hpp"
+#include "Serializable.h"
+#include "Logger.hpp"
+
+namespace bftEngine::bcst::impl {
+
+using RVBGroupId = uint64_t;
+using RVBId = uint64_t;
+using RVBIndex = uint64_t;
+
+// A Range Validation Tree (RVT)
+//
+// RVT is used to confirm that blocks collected from any single source (part of consensus network) during State Transfer
+// are valid and true. This is done by validating digest of certain blocks at frequent interval (RVB) against the tree.
+// RVT would be updated during checkpointing and also in the context of pruning. New nodes would be added to RVT during
+// checkpointing where as existing nodes would be deleted when pruning on old blocks begins.
+// Comment: this data strucutre is easy to be used to validate any type of data, not only block digests.
+//
+// Terms used throughout code -
+// 1. RVT_K = Maximum number of children any node can have
+// 2. Fetch range = Not all blocks will be validated. Only max block from given fetch range will be validated.
+// 3. RVB Id = Any block id in multiple of fetch range size
+// 4. RVB Index = RVBId / fetch range size
+// 5. RVB GroupIndex = Minimum of RVBIndex of all childrens
+//
+// Things to remember -
+// 1. Tree does not store RVB nodes.
+// 2. Only blocks at specific interval are validated to improve replica recovery time.
+// 3. Each node in tree is represented as NodeInfo having type.
+// 4. NodeVal is stored in form of CryptoPP::Integer.
+
+class RangeValidationTree {
+  // The next friend declerations are used strictly for testing
+  friend class BcStTestDelegator;
+
+  using NodeVal_t = CryptoPP::Integer;
+
+ public:
+  /////////////////////////// API /////////////////////////////////////
+  RangeValidationTree(const logging::Logger& logger, uint32_t RVT_K, uint32_t fetch_range_size, size_t value_size = 32);
+  ~RangeValidationTree() = default;
+  RangeValidationTree(const RangeValidationTree&) = delete;
+  RangeValidationTree& operator=(RangeValidationTree&) = delete;
+
+  // These API deals with range validation block ids
+  void addNode(const RVBId id, const char* data, size_t data_size);
+  void removeNode(const RVBId id, const char* data, size_t data_size);
+  // Return complete tree along with metadata in serialized format
+  std::ostringstream getSerializedRvbData() const;
+  // Initialize metadata & build tree by deserializing input stream
+  // If function fails, tree reset to null and returns false.
+  bool setSerializedRvbData(std::istringstream& iss);
+  // Returns RVB group ids in ascending order. In case of error, returns empty vector.
+  std::vector<RVBGroupId> getRvbGroupIds(RVBId start_block_id, RVBId end_block_id) const;
+  // Returns min actual and max actual RVB ids in ascending order. In case of error, returns empty vector.
```

Comment:
Please review all API functions header comments again:
1. Fix incorrect comments like in this line.
2. explain what happens in case of failure (and return value != void)